src/treelearner/serial_tree_learner.cpp

```diff
@@ -743,28 +747,24 @@ void SerialTreeLearner::RenewTreeOutput(Tree* tree, const ObjectiveFunction* obj
 }
 
 void SerialTreeLearner::ComputeBestSplitForFeature(
-    FeatureHistogram *histogram_array_,
-    const std::unique_ptr<LeafSplits>& leaf_splits_, int feature_index, int real_fidx,
-    const std::vector<int8_t>& node_used_features, const int tid,
-    std::vector<SplitInfo>& best) {
-
+    FeatureHistogram* histogram_array_, int feature_index, int real_fidx,
+    bool is_feature_used, int num_data, const LeafSplits* leaf_splits,
+    SplitInfo* best_split) {
+  if (!is_feature_used) {
+    return;
+  }
   SplitInfo new_split;
-
   histogram_array_[feature_index].FindBestThreshold(
-      leaf_splits_->sum_gradients(), leaf_splits_->sum_hessians(),
-      leaf_splits_->num_data_in_leaf(),
-      constraints_per_leaf_[leaf_splits_->LeafIndex()], &new_split);
-
+      leaf_splits->sum_gradients(), leaf_splits->sum_hessians(), num_data,
+      constraints_->Get(leaf_splits->leaf_index()), &new_split);
   new_split.feature = real_fidx;
-
   if (cegb_ != nullptr) {
     new_split.gain -=
-        cegb_->DetlaGain(feature_index, real_fidx, leaf_splits_->LeafIndex(),
-                         leaf_splits_->num_data_in_leaf(), new_split);
+        cegb_->DetlaGain(feature_index, real_fidx, leaf_splits->leaf_index(),
+                         num_data, new_split);
   }
-
-  if (new_split > best[tid] && node_used_features[feature_index]) {
-    best[tid] = new_split;
+  if (new_split > *best_split && is_feature_used) {
```

Comment:
I don't think there is a need to check "&& is_feature_used", because it is already checked above, line 753:

 if (!is_feature_used) {
    return;
  }